galley/pkg/ca/ca.go

```diff
@@ -0,0 +1,213 @@
+// limitations under the License.
+package ca
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+
+	"istio.io/istio/pkg/log"
+	"istio.io/istio/security/pkg/k8s/controller"
+	istioca "istio.io/istio/security/pkg/pki/ca"
+	"istio.io/istio/security/pkg/pki/util"
+)
+
+var (
+	// certChainID is the ID/name for the certificate chain file.
+	certChainID = "cert-chain.pem"
+	// privateKeyID is the ID/name for the private key file.
+	privateKeyID = "key.pem"
+	// rootCertID is the ID/name for the CA root certificate file.
+	rootCertID = "root-cert.pem"
+
+	galleyServiceAccountName = "istio-galley-service-account"
+
+	galleySecret = "istio." + galleyServiceAccountName
+)
+
+//GenerateSelfSignedCA generated self signed ca if istio.istio-galley-service-account does not exist
+func GenerateSelfSignedCA(caArgs *Args, workloadCertArgs *WorkloadCertArgs, kubeConfig, namespace, svcName string) error {
+
+	cs := createClientset(kubeConfig)
+	if !secretIsAlreadyExists(cs, namespace) {
+		err := createGalleySecret(cs, caArgs, workloadCertArgs, namespace, svcName)
+		if err != nil {
+			return fmt.Errorf("generate secret failed: %v", err)
+		}
+	} else {
+		return generateCAFilesAndWatchSecret(cs, caArgs, workloadCertArgs, namespace, svcName)
+	}
+	return nil
+}
+
+func createGalleySecret(clientset *kubernetes.Clientset, caArgs *Args, workloadCertArgs *WorkloadCertArgs,
+	namespace, svcName string) (err error) {
+	options := util.CertOptions{
+		TTL:          caArgs.RequestedCertTTL,
+		Org:          caArgs.Org,
+		IsCA:         caArgs.ForCA,
+		IsSelfSigned: true,
+		RSAKeySize:   caArgs.RSAKeySize,
+	}
+
+	pemCert, pemKey, ckErr := util.GenCertKeyFromOptions(options)
+	if ckErr != nil {
+		return fmt.Errorf("unable to generate CA cert and key for self-signed CA (%v)", ckErr)
+	}
+
+	caKeyCertBundle, err := util.NewVerifiedKeyCertBundleFromPem(pemCert, pemKey, nil, pemCert)
+	if err != nil {
+		return fmt.Errorf("failed to create CA KeyCertBundle (%v)", err)
+	}
+
+	err = createCAFiles(caArgs, pemCert, pemKey, pemCert)
+	if err != nil {
+		return err
+	}
+
+	return watchSecret(clientset, caArgs, workloadCertArgs,
+		caKeyCertBundle, namespace, svcName)
+
+}
+
+func watchSecret(clientset *kubernetes.Clientset, caArgs *Args, workloadCertArgs *WorkloadCertArgs,
+	caKeyCertBundle util.KeyCertBundle, namespace, svcName string) error {
+	istioCA, err := istioca.NewIstioCA(&istioca.IstioCAOptions{
+		CertTTL:       caArgs.CertTTL,
+		MaxCertTTL:    caArgs.MaxCertTTL,
+		KeyCertBundle: caKeyCertBundle,
+	})
+	if err != nil {
+		return err
+	}
+
+	webhooks := make(map[string]controller.DNSNameEntry)
+	webhooks[galleyServiceAccountName] = controller.DNSNameEntry{
+		ServiceName: svcName,
+		Namespace:   namespace,
+	}
+
+	sc, err := controller.NewSecretController(istioCA,
```

Comment:
I'm not sure if IstioCA package is intended to be used by other packages. Thoughts @wattli?